test/java/seedu/addressbook/commands/AddCommandTest.java

```diff
@@ -0,0 +1,159 @@
+                    Address.EXAMPLE, false, tags);
+        }
+    }
+    
+    @Test
+    public void addCommand_invalidEmail_throws() {
+        Set<String> tags = new HashSet<>();
+        final String[] invalidEmails = { "", " ", "def.com", "@", "@def", "@def.com", "abc@" };
+        for (String email : invalidEmails) {
+            assertConstructingInvalidAddCmdThrows(Name.EXAMPLE, Phone.EXAMPLE, false, email, false,
+                    Address.EXAMPLE, false, tags);
+        }
+    }
+    
+    @Test
+    public void addCommand_invalidAddress_throws() {
+        Set<String> tags = new HashSet<>();
+        final String[] invalidAddresses = { "", " " };
+        for (String address : invalidAddresses) {
+            assertConstructingInvalidAddCmdThrows(Name.EXAMPLE, Phone.EXAMPLE, false, Email.EXAMPLE, false,
+                    address, false, tags);
+        }
+    }
+    
+    @Test
+    public void addCommand_invalidTags_throws() {
+        final String[][] invalidTags = { { "" }, { " " }, { "'" }, { "validTag", "" }, { "", " " } };
+        for (String[] tags : invalidTags) {
+            Set<String> tagsToAdd = new HashSet<>(Arrays.asList(tags));
+            assertConstructingInvalidAddCmdThrows(Name.EXAMPLE, Phone.EXAMPLE, false, Email.EXAMPLE, false,
+                    Address.EXAMPLE, false, tagsToAdd);
+        }
+    }
+    
+    /**
+     * Asserts that attempting to construct an add command with the supplied
+     * invalid data throws an IllegalValueException
+     */
+    private void assertConstructingInvalidAddCmdThrows(String name, String phone, boolean isPhonePrivate,
+            String email, boolean isEmailPrivate, String address, boolean isAddressPrivate,
+            Set<String> tags) {
+        try {
+            new AddCommand(name, phone, isPhonePrivate, email, isEmailPrivate, address, isAddressPrivate,
+                    tags);
+        } catch (IllegalValueException e) {
+            return;
+        }
+        String error = String.format(
+                "An add command was successfully constructed with invalid input: %s %s %s %s %s %s %s %s",
+                name, phone, isPhonePrivate, email, isEmailPrivate, address, isAddressPrivate, tags);
+        fail(error);
+    }
+    
+    @Test
+    public void addCommand_validData_correctlyConstructed() throws IllegalValueException {
+        Set<String> tags = new HashSet<>();
+        AddCommand cmd = new AddCommand(Name.EXAMPLE, Phone.EXAMPLE, true, Email.EXAMPLE, false,
+                Address.EXAMPLE, true, tags); // should never throw
+        ReadOnlyPerson p = cmd.getPerson();
+        
+        assertEquals(p.getName().fullName, Name.EXAMPLE);
+        assertEquals(p.getPhone().value, Phone.EXAMPLE);
+        assertTrue(p.getPhone().isPrivate());
+        assertEquals(p.getEmail().value, Email.EXAMPLE);
+        assertFalse(p.getEmail().isPrivate());
+        assertEquals(p.getAddress().value, Address.EXAMPLE);
```

Comment:
Take note of the argument order of [`assertEquals`](http://junit.sourceforge.net/javadoc/org/junit/Assert.html#assertEquals(java.lang.Object,%20java.lang.Object)) -- The first argument is the expected value and the second argument is the actual value.